### PR TITLE
COP-12018 Implemented can user update targets

### PR DESCRIPTION
**Pull request by FrancisBry**

### CODE REVIEW

File: client/src/pages/targets/TargetListCard.test.jsx [hunk 1]
```diff
@@ -12,12 +13,22 @@ import mockTargetByVessel from './__fixtures__/mockTargetByVessel.json';
...
+  );
 
   const setTasksTabValue = (value, componentProps) => {
     return (
-      <TaskSelectedTabContext.Provider value={value}>
-        Ï<TargetListCard {...componentProps} />
-      </TaskSelectedTabContext.Provider>
+      <MockPermissionContext>
+        <TaskSelectedTabContext.Provider value={value}>
+          Ï<TargetListCard {...componentProps} />
```
Comment (charles474) [inline client/src/pages/targets/TargetListCard.test.jsx:29]: Typo error (you've inherited this by rebasing with Shakir's branch)

`Ï`

File: client/src/pages/targets/TasksTab.test.jsx [hunk 2]
```diff
@@ -61,11 +67,20 @@ describe('TasksTab', () => {
...
+  );
+
   const setTasksTabValue = (value, tabProps) => {
     return (
-      <TaskSelectedTabContext.Provider value={value}>
-        Ï<TasksTab {...tabProps} />
-      </TaskSelectedTabContext.Provider>
+      <MockPermissionContext>
+        <TaskSelectedTabContext.Provider value={value}>
+          Ï<TasksTab {...tabProps} />
```
Comment (charles474) [inline client/src/pages/targets/TasksTab.test.jsx:81]: Typo error `Ï`

File: client/src/pages/targets/components/Filter/Filter.jsx [hunk 3]
```diff
@@ -47,46 +48,49 @@ const Filter = ({
...
-    <>
-      <div className="cop-filters-header">
-        <h2 className="govuk-heading-s">{t(`pages.filters.title`)}</h2>
-      </div>
-      <div className="cop-filters-controls">
-        <div className="govuk-button-group">
-          <div className="govuk-!-padding-top-1">
-            <Button
-              className="govuk-!-margin-bottom-1"
+  if (permission) {
```
Comment (charles474) [inline client/src/pages/targets/components/Filter/Filter.jsx:51]: I would normally avoid rebasing/merging a branch which has not been merged into master as the code here, 've already asked Shaks to modify it.

**Review by charles474 — COMMENTED** [review]
Might be worth rebasing/merging with the `bluetargeting-dev` branch once his PR is merged then I'll go over it again. Feels like I'm looking at the same PR for the fourth time 😁.

File: client/src/pages/targets/TasksTab.test.jsx [hunk 4]
```diff
@@ -122,7 +123,7 @@ describe('TasksTab', () => {
     expect(screen.queryByText('View details')).not.toBeInTheDocument();
   });
 
-  it('should display options for an acknowledged task', async () => {
+  it('should display options for an confirm task receipt', async () => {
```
Comment (charles474) [inline client/src/pages/targets/TasksTab.test.jsx:126]: `should display options for an acknowledged target`? This test is still RoRo V1

File: client/src/pages/targets/components/TargetListItemOptions.jsx [hunk 5]
```diff
@@ -32,8 +33,9 @@ const ReceiptButton = ({ selectedTab, targetStatus, onClick }) => {
   return null;
 };
 
-const AcceptButton = ({ selectedTab, targetStatus, onClick }) => {
+const AcceptButton = ({ selectedTab, targetStatus, onClick, permission }) => {
```
Comment (charles474) [inline client/src/pages/targets/components/TargetListItemOptions.jsx:36]: Missing tests to test that these buttons render (in isolation to the rest of the UI). The reason why this component didn't really have a test file was due to duplicated tests which test the same functionaility.

File: client/src/pages/targets/v2/TargetsTab.test.jsx [hunk 6]
```diff
@@ -125,6 +126,21 @@ describe('TargetsTab', () => {
     expect(screen.queryByText('View details')).not.toBeInTheDocument();
   });
 
+  it('should not show view accept button if user does not have update permission', async () => {
```
Comment (charles474) [inline client/src/pages/targets/v2/TargetsTab.test.jsx:129]: Slight typo, the word `view` should not be in that sentence.

`should not show the accept button ...`

File: client/src/pages/targets/v2/TargetsTab.test.jsx [hunk 7]
```diff
@@ -167,6 +183,21 @@ describe('TargetsTab', () => {
     expect(screen.queryByText('Accept')).not.toBeInTheDocument();
   });
 
+  it('should not show view confirm receipt button if user does not have update permission', async () => {
```
Comment (charles474) [inline client/src/pages/targets/v2/TargetsTab.test.jsx:186]: Slight typo, the word `view` should not be in that sentence.

`should not show the confirm receipt button ...`

File: client/src/pages/targets/v2/TargetsTab.test.jsx [hunk 8]
```diff
@@ -235,6 +266,21 @@ describe('TargetsTab', () => {
     expect(localStorage.getItem(CACHED_TAB)).toEqual(TASK_STATUS.ASSIGNED);
   });
 
+  it('should not show view capture button if user does not have update permission', async () => {
```
Comment (charles474) [inline client/src/pages/targets/v2/TargetsTab.test.jsx:269]: Slight typo, the word `view` should not be in that sentence.

`should not show the capture button ...`

File: client/src/pages/targets/v2/TargetsTab.test.jsx [hunk 9]
```diff
@@ -259,6 +305,21 @@ describe('TargetsTab', () => {
     );
   });
 
+  it('should not show view record outcome button if user does not have update permission', async () => {
```
Comment (charles474) [inline client/src/pages/targets/v2/TargetsTab.test.jsx:308]: Slight typo, the word `view` should not be in that sentence.

`should not show the record outcome button ...`

File: client/src/pages/targets/v2/TargetsTab.test.jsx [hunk 10]
```diff
@@ -301,6 +362,21 @@ describe('TargetsTab', () => {
     expect(naviURLs[0]).toEqual(`/roro/targets/${targetId}`);
   });
 
+  it('should not show view undo capture button if user does not have update permission', async () => {
```
Comment (charles474) [inline client/src/pages/targets/v2/TargetsTab.test.jsx:365]: Slight typo, the word `view` should not be in that sentence.

`should not show the undo capture button ...`

File: client/src/pages/tasks/components/TaskListItem.jsx [hunk 11]
```diff
@@ -8,6 +8,7 @@ import { Link, useNavigation } from 'react-navi';
 import { isOverDue } from './utils';
 import { useAxios } from '../../../utils/hooks';
 import './__styles__/TaskListItem.scss';
+import { PermissionContext } from '../../../utils/PermissionContext';
```
Comment (charles474) [inline client/src/pages/tasks/components/TaskListItem.jsx:11]: This line needs to be the last import `import './__styles__/TaskListItem.scss';`

File: client/src/pages/tasks/components/TaskListItem.jsx [hunk 12]
```diff
@@ -99,16 +114,7 @@ const TaskListItem = ({ id, due, name, assignee, businessKey, taskType, pageType
...
-                <button
-                  type="submit"
-                  id="actionButton"
-                  className="govuk-button"
-                  onClick={handleClaim}
-                >
-                  Action
-                </button>
-              </div>
+              <div className="govuk-grid-column-one-third claim-task">{canActionTask()}</div>
```
Comment (charles474) [inline client/src/pages/tasks/components/TaskListItem.jsx:117]: `canActionTask` suggests that this is a conditional.

what you can do is create a sub-component for the action task button and within that component, do the check and return the button as the default but a guar clause above it.


```
  const ActionTaskButton = ({ onClick }) => {
    if (!canUpdate) {
      return null;
    }
    return (
      <button type="submit" id="actionButton" className="govuk-button" onClick={onClick}>
        Action
      </button>
    );
  };


....



<div className="govuk-grid-column-one-third claim-task">
  <ActionTaskButton onClick={onClick} />
</div>
```

File: client/src/pages/tasks/components/TaskListItem.jsx [hunk 13]
```diff
@@ -64,6 +69,16 @@ const TaskListItem = ({ id, due, name, assignee, businessKey, taskType, pageType
       </button>
     );
   };
+  const canActionTask = () => {
```
Comment (charles474) [inline client/src/pages/tasks/components/TaskListItem.jsx:72]: Same as comment below https://github.com/UKHomeOffice/cop-ui/pull/601/files#r1008054041

File: client/src/pages/targets/TasksTab.test.jsx (hunk 4 above)
Comment (FrancisBry) [reply client/src/pages/targets/TasksTab.test.jsx]: Reverted it

File: client/src/pages/targets/components/TargetListItemOptions.jsx (hunk 5 above)
Comment (FrancisBry) [reply client/src/pages/targets/components/TargetListItemOptions.jsx]: I added tests for all these

File: client/src/pages/targets/v2/TargetsTab.test.jsx (hunk 6 above)
Comment (FrancisBry) [reply client/src/pages/targets/v2/TargetsTab.test.jsx]: The dangers of copy and paste. I fixed these

**Review by charles474 — APPROVED** [review]
Thanks for making the changes.

**Pull request merged**